greenpithumb/sleep_windows.py

```diff
@@ -0,0 +1,84 @@
+"""Parses sleep window strings."""
+
+import datetime
+import re
+
+
+class Error(Exception):
+    pass
+
+
+class InvalidWindowFormatError(Error):
+    pass
+
+
+class HypersomniaError(Error):
+    pass
+
+
+def parse(sleep_windows):
+    """Converts sleep windows from strings to datetime objects.
+
+    Args:
+        sleep_windows: A list of sleep window strings.Each sleep window should
+            be in 24-hour HH:MM-HH:MM format.
+
+    Returns:
+        A list of 2-tuples, each representing a sleep window. Tuple items
+        are datetime.time objects.
+    """
+    awake_minutes = [1] * 60 * 24
+    sleep_minutes = set()
+    sleep_windows_datetime = []
+    for sleep_window in sleep_windows:
+        try:
+            window_match = re.match(r'(\d\d):(\d\d)-(\d\d):(\d\d)',
+                                    sleep_window)
+        except TypeError:
+            raise InvalidWindowFormatError(
+                'Time windows must be in string format.')
+        if not window_match:
+            raise InvalidWindowFormatError(
+                'Sleep windows must be in 24-hour HH:MM-HH:MM format')
+        try:
+            sleep_time = datetime.time(
+                int(window_match.group(1)), int(window_match.group(2)))
+            wake_time = datetime.time(
+                int(window_match.group(3)), int(window_match.group(4)))
+        except (ValueError, TypeError):
+            raise InvalidWindowFormatError(
+                'Sleep windows must be in 24-hour HH:MM-HH:MM format')
+        if sleep_time == wake_time:
+            raise HypersomniaError(
+                'Sleep windows cannot cover the entire 24-hour range.')
+        sleep_windows_datetime.append((sleep_time, wake_time))
+        sleep_minutes.update(_sleep_minutes(sleep_time, wake_time))
+    for minute in sleep_minutes:
+        awake_minutes[minute] = 0
+    if not sum(awake_minutes):
+        raise HypersomniaError(
+            'Sleep windows cannot cover the entire 24-hour range.')
+
+    return sleep_windows_datetime
+
+
+def _sleep_minutes(sleep_time, wake_time):
+    """Calculates the individual minutes in given sleep window.
+
+    Args:
+        sleep_time: Sleep time as a datetime object.
+        wake_time: Wake time as a datetime object.
+
+    Returns:
+        A set containing all of the minutes in the sleep window, from 0 to 1439.
```

Comment:
(maybe a moo point per my other comment, but): This could use a bit more detail. Saying "from 0 to 1439" gives the impression that the return set always includes all of those integers.

Also the number `1439` is confusing because it's not immediately recognizable as 24 * 60 - 1, so for things like that, suggest including the calculation too so it doesn't seem like a magic number.